xinference/core/tests/test_restful_api.py

```diff
@@ -691,6 +691,67 @@ def test_restful_api_for_qwen_tool_calls(setup, model_format, quantization):
     response_data = response.json()
     assert len(response_data["data"]) == 1
 
+    url = f"{endpoint}/v1/chat/completions"
+    payload = {
+        "model": model_uid_res,
+        "messages": [
+            {
+                "role": "user",
+                "content": "谁是周杰伦？",
+            },
+        ],
+        "tools": [],
+        "max_tokens": 2048,
+        "temperature": 0,
+    }
+    response = requests.post(url, json=payload)
+    completion = response.json()
+    assert "stop" == completion["choices"][0]["finish_reason"]
+
+    tools = [
+        {
+            "type": "function",
+            "function": {
+                "name": "google_search",
+                "description": "谷歌搜索是一个通用搜索引擎，搜索周杰伦。",
+                "parameters": {},
+            },
+        },
+        {
+            "type": "function",
+            "function": {
+                "name": "image_gen",
+                "description": "文生图是一个AI绘画（图像生成）服务，画个周杰伦。",
+            },
+        },
+    ]
+
+    url = f"{endpoint}/v1/chat/completions"
+    payload = {
+        "model": model_uid_res,
+        "messages": [
+            {
+                "role": "user",
+                "content": "谁是周杰伦？",
+            },
+        ],
+        "tools": tools,
+        "max_tokens": 2048,
+        "temperature": 0,
+    }
+    response = requests.post(url, json=payload)
+    completion = response.json()
+    assert "content" in completion["choices"][0]["message"]
+    assert "tool_calls" == completion["choices"][0]["finish_reason"]
+    assert (
+        "google_search"
+        == completion["choices"][0]["message"]["tool_calls"][0]["function"]["name"]
+    )
+    arguments = completion["choices"][0]["message"]["tool_calls"][0]["function"][
+        "arguments"
+    ]
+    assert not json.loads(arguments)
```

Comment:
I tested and change back to 7b, 4b will failure earlier, and for 7b it will fail at this line.
```
>       assert not json.loads(arguments)
E       assert not {'query': '周杰伦'}
E        +  where {'query': '周杰伦'} = <function loads at 0x7fd1171cfc40>('{"query": "\\u5468\\u6770\\u4f26"}')
E        +    where <function loads at 0x7fd1171cfc40> = json.loads
```

Reply:
I updated the ut and tested the latest 4b and 7b qwen 1.5:
- 4b
  - Can't pass the false tool call check (the model should not call any tools because there are no matched tools), so I removed this check.
  - The output varies between torch and ggufv2 models, so I removed this check.
- 7b
  - Seems does not support tool calls. All the tool calls are failed.
  - torch: Can't tool call with or without tool parameters.
  - ggufv2 Q5_K_M: Output 9!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Reply:
I think the qwen 1.5 model has problems making tool calls. My env:

pip Name: transformers
Version: 4.39.1

Name: llama_cpp_python
Version: 0.2.64

Reply:
How about just using 7b? Maybe 4b is not good enough.

Reply:
I have tested 7b, as shown above, it does not support tool calls. All the tests are failed. 4b is OK.

Reply:
I cleaned my local cache and downloaded the latest model to test.

Reply:
The pytorch version can pass, but failed for gguf, is this PR ready for review?